### PR TITLE
ToPythonTransformer + DataTransformer unit tests

### PR DESCRIPTION
**Pull request by Radinyn**

### CODE REVIEW

File: protostar/utils/data_transformer.py [hunk 1]
```diff
@@ -9,35 +9,76 @@
...
+            raise DataTransformerException(str(ex)) from ex
+
+    return transform
+
+
+def to_python_events_transformer(
+    contract_abi: AbiType, event_name: str
+) -> ToPythonTransformer:
+    structure_transformer = CairoSerializer(identifier_manager_from_abi(contract_abi))
+    event_abi_item = find_abi_item(contract_abi, event_name)
```
Comment (mkaput) [inline protostar/utils/data_transformer.py:74]:
```suggestion
    event_abi_item = find_abi_item(contract_abi, event_name)
    structure_transformer = CairoSerializer(identifier_manager_from_abi(contract_abi))
```

to be aligned with `to_python_transformer` impl. and please do the same in `from_python_transformer`

File: protostar/utils/data_transformer_test.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        "type": "event",
+        "from_address": "0",
+    },
+    {
+        "data": [{"name": "input", "type": "(felt, felt)"}],
+        "name": "tuple_event",
+        "type": "event",
+        "from_address": "0",
+    },
+]
```
Comment (kasperski95) [inline protostar/utils/data_transformer_test.py:91]: Can you create a test utility function that generates the ABI? It would make the test suite more readable and maintainable.

 e.g.
```python
abi = create_abi(
"""\
@constructor
func 

# ...
"""
)
```

Also, you split ABI/PYTHON_INPUT/CAIRO_INPUT into triplets grouped by the input type.
arg, felt_arg, uint256, etc.

**Review by mkaput — COMMENTED** [review]
@Radinyn could you please outline to me how are you planning to use this for input validation? I know that you will depend on the exceptions being raised here, but I would love to see the incoming code :)

File: protostar/utils/data_transformer_test.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    [
+        ["constructor", "inputs", VALID_CONSTRUCTOR_PYTHON_INPUT],
+        ["increase_balance", "outputs", {}],
+        ["get_balance", "outputs", {"res": 0x42}],
+    ],
+)
+def test_from_python_pass(
+    name: str, mode: Literal["inputs", "outputs"], data: PythonData
+):
+    from_python_transformer(TEST_ABI, name, mode)(data)
```
Comment (kasperski95) [inline protostar/utils/data_transformer_test.py:156]: You don't check the output of the transformer

Consider:
```python
cairo_input = from_python(python_input)
assert_cairo_input_is_list(cairo_input)
python_input_2 = to_python(cairo_input)
assert python_input == python_input_2
```

File: protostar/utils/data_transformer_test.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
+from typing_extensions import Literal
```
Comment (kasperski95) [inline protostar/utils/data_transformer_test.py:1]: I'd say this is an integration test because it relies on Starknet.py

File: protostar/utils/data_transformer.py (hunk 1 above)
Comment (Radinyn) [reply protostar/utils/data_transformer.py:74]: DataTransformer is going to raise an exception when output is invalid which will be later caught by the validator. I know using a transformer for validation isn't particularly clean but I don't think we have much options if we want to delegate it to starknet.py. Nevertheless afaik we wanted to add this functionality to the DataTransformer either way.

Comment (mkaput) [reply protostar/utils/data_transformer.py:74]: OK, would it make sense to do wrapping of ValueError/TypeError exceptions into DataTransformerException in validation code instead of here? In order to make such shenanigans more use-case specific.

Comment (Radinyn) [reply protostar/utils/data_transformer.py:74]: I don't think we would ever want Protostar to crash on DataTransformer fail, this way we can finish cleanly.

Comment (mkaput) [reply protostar/utils/data_transformer.py:74]: hmm, right. please reorder statements then and resolve this discussion :)

File: protostar/utils/data_transformer_test.py (hunk 4 above)
Comment (mkaput) [reply protostar/utils/data_transformer_test.py:1]: But we are not testing integration of our code swatches; instead we test single functions which uses Starknet.py as an internal dependency. 🤔

Comment (Radinyn) [reply protostar/utils/data_transformer_test.py]: I've decided to go with the integration tests approach, because it allows us to leverage ProtostarFixture.

File: protostar/utils/data_transformer_test.py (hunk 3 above)
Comment (Radinyn) [reply protostar/utils/data_transformer_test.py]: done

File: protostar/utils/data_transformer_test.py (hunk 2 above)
Comment (Radinyn) [reply protostar/utils/data_transformer_test.py]: done

File: tests/data/contracts.py [hunk 5]
```diff
@@ -59,3 +59,76 @@
...
+
+DATA_TRANSFORMER_OUTPUT_FELT_CONTRACT = """%lang starknet
+
+@external
+func output_felt() -> (res: felt):
+    return (42)
+end
+"""
+
+DATA_TRANSFORMER_EVENT_CONTRACT = """%lang starknet
```
Comment (kasperski95) [inline tests/data/contracts.py:129]: Consider making names more generic, or moving contracts to the data transformer test suite.

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -182,3 +184,17 @@ def create_protostar_project():
...
 
     return create_protostar_project
+
+
+@pytest.fixture(name="get_abi_from_contract", scope="module")
+def get_abi_from_contract_fixture(create_protostar_project) -> Callable[[str], AbiType]:
+    def get_abi_from_contract(contract_source_code: str) -> AbiType:
+        with create_protostar_project() as protostar:
+            with open("src/main.cairo", "w") as f:
+                f.write(contract_source_code)
```
Comment (kasperski95) [inline tests/integration/conftest.py:194]:
```suggestion
            protostar.create_files({"src/main.cairo", contract_source_code})
```

File: tests/integration/data_transformer/data_transformer_test.py [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+        #     "inputs",
+        #     ["very very very very long string"],
+        # ],
+        # AFTER STARKNET.py ADDS EXCEPTION ON TOO MANY VALUES
+        # [
+        #     DATA_TRANSFORMER_FELT_CONSTRUCTOR_CONTRACT,
+        #     "constructor",
+        #     "inputs",
+        #     [0x2342, 0x24],
+        # ],
```
Comment (kasperski95) [inline tests/integration/data_transformer/data_transformer_test.py:259]: - Can you add links to Starknet.py issues?
- Consider using `@pytest.mark.skip` (check `gateway_facade_test.py::test_call_to_with_incorrect_args`). 
- WHY DO YOU WRITE COMMENTS IN **UPPERCASE**?

File: tests/integration/data_transformer/data_transformer_contract.cairo [hunk 8]
```diff
@@ -0,0 +1,124 @@
+
```
Comment (kasperski95) [inline tests/integration/data_transformer/data_transformer_contract.cairo]: Is this file still needed?

File: tests/data/contracts.py (hunk 5 above)
Comment (Radinyn) [reply tests/data/contracts.py]: moved

File: tests/integration/conftest.py (hunk 6 above)
Comment (Radinyn) [reply tests/integration/conftest.py]: done

File: tests/integration/data_transformer/data_transformer_test.py (hunk 7 above)
Comment (Radinyn) [reply tests/integration/data_transformer/data_transformer_test.py]: done

**Review by kasperski95 — APPROVED** [review]

File: tests/integration/data_transformer/contracts.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
+DATA_TRANSFORMER_UINT256_CONTRACT = """%lang starknet
```
Comment (mkaput) [inline tests/integration/data_transformer/contracts.py:1]:
```suggestion
DATA_TRANSFORMER_UINT256_CONTRACT = """
%lang starknet
```

does this work? if not, can you put `.strip()` there? or `"""\`

**Review by mkaput — APPROVED** [review]
